src/Evolution/Systems/Cce/BoundaryData.hpp

```diff
@@ -221,4 +333,218 @@ void dyads(
         up_dyad) noexcept;
 
+/*!
+ * \brief Compute the \f$\beta\f$ (lapse) function for the CCE Bondi-like
+ * metric.
+ *
+ * \details The Bondi-like metric has \f$g^{u r} = - e^{2 \beta}\f$, and the
+ * value of \f$\beta\f$ is obtained from the intermediate null metric by (see
+ * equation (51) of \cite Barkett2019uae) using:
+ *
+ * \f[
+ * \beta = -\frac{1}{2} \ln r_{, \lamdba}
+ * \f]
+ */
+void beta_worldtube_data(
+    gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 0>>*> beta,
+    const tnsr::a<DataVector, 3, Frame::RadialNull>& d_bondi_r) noexcept;
+
+/*!
+ * \brief Compute the \f$U\f$ (shift) function for the CCE Bondi-like metric.
+ *
+ * \details The Bondi-like metric has \f$g^{r A} = -e^{-2 \beta} U^A\f$, and the
+ * spin-weighted vector \f$U = U^A q_A\f$. The value of \f$U^A\f$ can be
+ * computed from the intermediate null metric quantities (see equation (54) of
+ * \cite Barkett2019uae) using:
+ *
+ * \f[
+ * U = -( r_{,\lambda} g^{\lambda A} + r_{,B} g^{A B}) q_A / r_{,\lambda}
+ * \f]
+ *
+ */
+void bondi_u_worldtube_data(
+    gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 1>>*> bondi_u,
+    const tnsr::i<ComplexDataVector, 2, Frame::RadialNull>& down_dyad,
+    const tnsr::a<DataVector, 3, Frame::RadialNull>& d_bondi_r,
+    const tnsr::AA<DataVector, 3, Frame::RadialNull>&
+        inverse_null_metric) noexcept;
+
+/*!
+ * \brief Compute the \f$W\f$ (mass aspect) function for the CCE Bondi-like
+ * metric.
+ *
+ * \details The Bondi-like metric has \f$g^{rr} = e^{-2 \beta}(1 + r W)\f$. The
+ * value of \f$W\f$ can be computed from the null metric quantities (see
+ * equation (55) of \cite Barkett2019uae) using:
+ *
+ * \f[
+ * W = \frac{1}{r} \left(-1
+ * + \frac{g^{\lambda \lambda} (r_{, \lambda})^2
+ * + 2 r_{, \lambda} \left(r_{,A} g^{\lambda A}
+ * - r_{,u}\right) + r_{, A} r_{, B} g^{A B}}{r_{, \lambda}}\right)
+ * \f]
+ */
+void bondi_w_worldtube_data(
+    gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 0>>*> bondi_w,
+    const tnsr::a<DataVector, 3, Frame::RadialNull>& d_bondi_r,
+    const tnsr::AA<DataVector, 3, Frame::RadialNull>& inverse_null_metric,
+    const Scalar<SpinWeighted<ComplexDataVector, 0>>& bondi_r) noexcept;
+
+/*!
+ * \brief Compute the \f$J\f$ (intuitively similar to the transverse-traceless
+ * part of the angular metric) function for the CCE Bondi-like metric.
+ *
+ * \details The Bondi-like metric has \f$J = \frac{1}{2 r^2} q^A q^B g_{A B}\f$.
+ * This expression holds both for the right-hand side in the Bondi coordinates
+ * and for the right-hand side in the intermediate null coordinates (see
+ * equation (45) of \cite Barkett2019uae).
+ */
+void bondi_j_worldtube_data(
+    gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 2>>*> bondi_j,
+    const tnsr::aa<DataVector, 3, Frame::RadialNull>& null_metric,
+    const Scalar<SpinWeighted<ComplexDataVector, 0>>& bondi_r,
+    const tnsr::I<ComplexDataVector, 2, Frame::RadialNull>& up_dyad) noexcept;
```

Comment:
just call `up_dyad` and `down_dyad` by `dyad` as the valence is encoded in the type